File: mobile/lib/models/devoir.dart
class Devoir {
  final int id;
  final String code;
  final String date;
  final double coeff;
  final int? typeDevoirId;
  final String? typeCode;
  final int? matiereId;
  final String? matiere;
  final String? matiereAbbr;
  final int? classeId;
  final String? classe;
  final String? classeAbbr;
  final int? niveauId;
  final String? niveau;
  final int? periodeId;
  final String? periode;

  const Devoir({
    required this.id,
    required this.code,
    required this.date,
    required this.coeff,
    this.typeDevoirId,
    this.typeCode,
    this.matiereId,
    this.matiere,
    this.matiereAbbr,
    this.classeId,
    this.classe,
    this.classeAbbr,
    this.niveauId,
    this.niveau,
    this.periodeId,
    this.periode,
  });

  factory Devoir.fromJson(Map<String, dynamic> json) => Devoir(
        id:          int.parse(json['id'].toString()),
        code:        json['code_devoir'] as String? ?? '',
        date:        json['date_devoir'] as String? ?? '',
        coeff:       double.tryParse(json['coeff_devoir'].toString()) ?? 1,
        typeDevoirId: json['type_devoir_id'] != null
            ? int.tryParse(json['type_devoir_id'].toString())
            : null,
        typeCode:    json['type_devoir']?['code_type_devoir'] as String?,
        matiereId:   json['matiere_id'] != null
            ? int.tryParse(json['matiere_id'].toString())
            : null,
        matiere:     json['matiere']?['libelle_matiere'] as String?,
        matiereAbbr: json['matiere']?['abbr_matiere'] as String?,
        classeId:    json['classe_id'] != null
            ? int.tryParse(json['classe_id'].toString())
            : null,
        classe:      json['classe']?['nom_classe'] as String?,
        classeAbbr:  json['classe']?['abbr_classe'] as String?,
        niveauId:    json['niveau_id'] != null
            ? int.tryParse(json['niveau_id'].toString())
            : null,
        niveau:      json['niveau']?['libelle_niveau'] as String?,
        periodeId:   json['periode_id'] != null
            ? int.tryParse(json['periode_id'].toString())
            : null,
        periode:     json['periode']?['libelle_periode'] as String?,
      );

  String get cible => classeAbbr ?? niveau ?? '—';
  bool get estNiveau => niveauId != null && classeId == null;
}

File: back/app/Models/Devoir.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Devoir extends Model
{
    use HasFactory;

    public $timestamps = false;

    protected $casts = ['coeff_devoir' => 'float'];

    protected $fillable = [
        'code_devoir',
        'date_devoir',
        'coeff_devoir',
        'type_devoir_id',
        'matiere_id',
        'classe_id',
        'niveau_id',
        'periode_id',
    ];

    public function typeDevoir()
    {
        return $this->belongsTo(TypeDevoir::class, 'type_devoir_id');
    }

    public function matiere()
    {
        return $this->belongsTo(Matiere::class);
    }

    public function classe()
    {
        return $this->belongsTo(Classe::class);
    }

    public function niveau()
    {
        return $this->belongsTo(Niveau::class);
    }

    public function periode()
    {
        return $this->belongsTo(Periodes::class, 'periode_id');
    }

    public function notes()
    {
        return $this->hasMany(Note::class);
    }
}
